# Лабораторная работа 4. Файлы и исключения

In [ ]:
Фамилия: Ляшко
Группа: ИНБ-б-о-26-2
Вариант: 6

## Задание

В файле data_v1.txt содержится последовательность символов, содержащая буквы латинского алфавита и десятичные цифры. Необходимо решить следующие задачи:
1. Найдите максимальное четное число в строке (под числом понимается последовательность цифр, ограниченная буквами).
2. Найдите буквенную подпоследовательность максимальной длины (то есть подпоследовательность не содержащую цифр). В ответе укажите ее длину.
3. Найдите символ в файле, который встречается чаще всего. Если таких символов несколько, то выведите все эти символы, в конце укажите одно число – сколько раз встречаются эти символы.

**В файле data_v2.txt найдите минимальное число кратное 3 (под
числом понимается непрерывная последовательность цифр)**

Индивидуальное задание

В файле data_v2.txt найдите минимальное число кратное 3 (под
числом понимается непрерывная последовательность цифр)


## Анализ задания, получение данных

Файл data_v1.txt расположен репозитории. Выполним копирование файла в локальное хранилище Google Colab, используя следующую команду:

In [ ]:
!wget https://raw.githubusercontent.com/enikolaev/IT-and-Programming/main/data/data_v2.txt


--2023-01-30 09:15:55--  https://raw.githubusercontent.com/enikolaev/IT-and-Programming/main/data/data_v1.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 100002 (98K) [text/plain]
Saving to: ‘data_v1.txt’

data_v1.txt         100%[===================>]  97.66K  --.-KB/s    in 0.002s  

2023-01-30 09:15:55 (47.5 MB/s) - ‘data_v1.txt’ saved [100002/100002]



Можно скопировать и поместить файл с данными любым удобным способом

Реализуем чтение всего файла в строку с использованием метоа `open()` и `read()`:

In [ ]:
f=open('data_v1.txt')
s=f.read().strip()
print(len(s))

100000


Обратите внимание на использование методв `strip()`

## Обработка символьной последовательности

Строки, подобно списка, являются коллекциями. Только в качестве элементов последовательности выступают символы. Для строк также используются функции нахождения длины `len()`, операция индексации `[]`

Например, можно посимвольно обходить строку, используя цикл `for` и индекс строки:

In [ ]:
# Вывод первых 10 символов строки с одновременным анализом буква это или цифра
for i in range(10):
  c = s[i]
  if c.isdigit():
    t = 'цифра'
  else:
    t='буква'
  print(c, t)

Z буква
S буква
t буква
M буква
P буква
i буква
7 цифра
c буква
s буква
i буква


## Решение задачи 1

Будем посимвольно обходимть строку и накапливать символы-цифры в строке `target`. Если цифровая последовательность прервется, то конвертируем `target` в число `int` и проверяем является ли число четным и не превысило ли оно максимум

In [ ]:
target=''
maxx=0
for c in s:
  if c.isdigit():
    target+=c
  else:
    if target!='':
      x = int(target)
      maxx=max(x, maxx)
      target=''

print('Максимальное четное число = %d' % maxx)


Максимальное четное число = 838478


## Решение задачи 2

Решение данной задачи осуществляется аналогично предыдущей. Можно даже вывести самуюдлинную буквенную строку

In [ ]:
target=''
maxx=0
maxs=''
for c in s:
  if c.isalpha():
    target+=c
  else:
    if len(target)>maxx:
      maxx=len(target)
      maxs=target
      target=''

print('Максимальная длина буквенной строки = %d' % maxx)
print('Сама строка:', maxs)

Максимальная длина буквенной строки = 1025
Сама строка: rpfunCMcMCfmnlNBbiCkCOPQsRYZmLjneFWZJjHqUxiTFGtMiKUTvaUkXBQnvGgpnDmUtFNyXZwWIdyqBWzLEqsTqoXCOkPaFQTuVxYeTftNLcEEjnRUqqAfeOmBHdcIDfyBpmvYjIivcZXMaflpfECMZKprJUSPIJtoolmBwWFmTiXtLsUgZrirYdtPUGEhsmAgGwqxoTcNlbtFmVApQAkcRiRFbvQxKwpdEVZhgUAymnZnRnJVmnDIvHmxtJkiICukqvSsMRrAHDBsAFDBWieGPhTcXNdetYzIdITVgvyiPqhPTKtixxggpYfCThpmcEqclpTTjnnHAXbpWOaOPlGcabgWvgiDDOhSWdTviUlNvTaoEVhoBfgrJkoJxtyDWXjVYYRokqkKPwRdkPkQYaVIoHCAfYPEyHgBsbUkYQyKNfgyWNJJsmtAoTPGrHKwUyJDvCEqgeRLTFdpqxaZFFGJrmkQuqbyadKxgWFsznrBtZyuubYwcmDAMqCqWXSUomuCMcFWWoxAWMJfdcYlDGMDSSCnYCUOVvgENGKZiTXKuLaLtMHTPjKmWwjwhfaePXiOPwZdvnAlPruuGwKhbRevjOHPVWIShguJJgKLaqUMJLwokvxJzYpvfOvfzRDHGQbnrWxESwCdsUzxyFWEYqRXpihePHiHXbMXBYyvIcUlaudLjXIPdXdMHrExnGgHeanzgtDSgSGOkxupzNQitVYWFXSNOpsdKMAveZQieNAqCMEXZzBdemldlsWCpiIMuQomlFEbZKcBwxAbNlFGPejoMQxgPHzXiskhadywliHoMmZlokYoauVlxmuHrqhwJHYpgzfUAqwPBAQPJNKlKhhYCRrLocFcIEorvZnBPIjlLVtMzafAoNJRufesTkbJcvviKEOSowgawXzWsFeUYgGsrdXZHbRGVQGvSczyAvdAeLjS

## Решение задачи 3

Для решения данной задачи будем использовать коллекцию словарь `dict()`. Данная коллекция содержит элемкнты в виде пар `ключ : значение`. Доступ к элементам осуществляется по ключу с использованием операции ндексации.

Заполним словарь d, который после прохода по строке должен принять вид:
d = {'A' : 10, 'B' : 127, '3' : 217, ... }
то есть в качестве ключа будем использовать символ, а вкачестве значения - количество таких символов в строке:

In [ ]:
# Создаем пустой словарь
d={}
# Проходим по строке и заполняем словарь
for c in s:
  d[c] = d.get(c, 0) + 1

# Выводим получившийся словарь:
print(d)

{'Z': 1623, 'S': 1613, 't': 1711, 'M': 1618, 'P': 1670, 'i': 1542, '7': 1643, 'c': 1562, 's': 1606, 'T': 1699, 'b': 1555, 'r': 1581, 'm': 1686, 'E': 1662, 'h': 1596, '8': 1654, 'f': 1609, 'B': 1647, 'K': 1616, 'a': 1586, 'F': 1623, 'O': 1580, 'w': 1582, 'o': 1609, '5': 1631, '1': 1580, 'W': 1595, '6': 1597, '0': 1651, 'I': 1604, 'u': 1606, 'e': 1566, 'p': 1644, 'D': 1644, 'A': 1638, 'Q': 1544, 'J': 1638, 'U': 1629, 'R': 1590, 'C': 1570, 'G': 1597, 'k': 1618, 'j': 1603, 'N': 1636, 'l': 1583, '2': 1600, 'q': 1614, 'X': 1619, 'H': 1565, '3': 1587, 'L': 1635, 'Y': 1647, '9': 1611, 'V': 1601, 'x': 1654, 'y': 1593, 'g': 1603, '4': 1599, 'v': 1655, 'd': 1599, 'z': 1615, 'n': 1566}


В получившемся словаре необходимо найти самый частотный символ. Для этого не будем обходить словарь, а просто воспользуемся встроенной функцией `max()`, которую применим к коллекции `values()` словаря

In [ ]:
max_val = max(d.values())
print(max_val)

1711


Выполним обход словаря и напечатаем ответ к данной задаче:

In [ ]:
for k in d:
  if d[k]==max_val:
    print(k, end=' ')
print(max_val)

t 1711


Для данного файла только один символ `t` встречается в файле максимальное `1711` количество раз

В данной работе были использованы операторы `if` (условный оператор), `for ... in range(...)` (оператор цикла с обходом диапазона). Применялись функции `print()`, `len()`, `max()`. В работе использовался тип данных `dict()`

In [ ]:
# Решение индивидуальной задачи
f = open('data_v2.txt').readline()
number = ''
numbers = []
for i in f:
	if i >= "0" and i <= "9":
		number = number + i
	else:
		if number != '':
			numbers.append(int(number))
			number = ''

kratniy_3 = []
for i in numbers:
	if i % 3 == 0:
		kratniy_3.append(i)

if len(kratniy_3) > 0:
	print(min(kratniy_3))



Ответы на контрольные вопросы:
1. Назначение, сходства и различия read(), readline(), readlines()
    
read()	Читает весь файл целиком (или n символов, если указать аргумент)
readline()	Читает одну строку (включая \n)
readlines()	Читает все строки файла

Сходства:

    Все три метода читают данные из текстового файла.
    Все они перемещают позицию считывания файла вперёд.
    Все возвращают строковые данные (или список строк).
    После достижения конца файла возвращают пустые значения ('' или []).

Различия:

    read() — читает всё сразу.

    readline() — читает по одной строке.

    readlines() — читает весь файл, но возвращает список строк.

2. Как прочитать многострочный файл без read(), readline(), readlines()

Файловый объект в Python является итерируемым, поэтому можно использовать цикл for:
```python

with open('file.txt', 'r') as f:
    for line in f:
        print(line.rstrip())
```
Также можно использовать list(f) для получения списка строк или f.read().splitlines(), но последнее уже использует read().
3. Задача закрытия файла

Закрытие файла:

    освобождает системные ресурсы (файловый дескриптор);

    гарантирует, что все буферизованные данные записаны на диск;

    предотвращает утечки ресурсов и возможные ошибки при дальнейшей работе;

    снимает блокировку файла (если она была).

В Python рекомендуется использовать конструкцию with open(...), которая автоматически закрывает файл.
4. Что такое позиция считывания файла?

Позиция считывания — это указатель на место в файле, с которого будет прочитан следующий символ или байт. При открытии файла в режиме чтения ('r') позиция изначально находится в начале файла (смещение 0).
5. Режим для записи с сохранением содержимого

Нужен режим дозаписи — 'a' (append). Данные записываются в конец файла, существующее содержимое остаётся нетронутым.
```python

with open('file.txt', 'a') as f:
    f.write('новая строка\n')
```
6. Что такое исключение?

Исключение — это событие, возникающее во время выполнения программы, которое нарушает нормальный ход её выполнения. Оно сигнализирует об ошибке или нештатной ситуации (например, деление на ноль, отсутствие файла, неверный тип данных). Исключения можно перехватывать и обрабатывать с помощью try/except.
7. Что происходит, если исключение не обрабатывается?

Если исключение вызвано, но не обработано инструкцией try/except, программа аварийно завершается, а в консоль выводится трассировка стека (traceback) с информацией об ошибке и месте её возникновения.
8. Исключение при открытии несуществующего файла

При попытке открыть несуществующий файл в режиме чтения возникает исключение:
```text

FileNotFoundError

(подкласс OSError).
```    
9. Исключение при float() для нечисловой строки

При конвертации нечислового строкового значения в число возникает исключение:
```text

ValueError
```
Пример:
```python

float('abc')  # ValueError: could not convert string to float: 'abc'
```
10. Что происходит при открытии несуществующего файла в режиме дозаписи?

Если файл не существует и программа открывает его в режиме дозаписи ('a'), файл будет создан автоматически, и запись начнётся с начала (с нулевой позиции). Исключения не возникает.
```python

with open('new_file.txt', 'a') as f:
    f.write('данные\n')  # файл будет создан
```
